reference-api/unified_handler.go

```diff
@@ -34,75 +44,109 @@ func (rec *responseRecorder) WriteHeader(code int) {
-			" your application repository"
-		_, err := fmt.Fprintln(w, errMsg)
-		if err != nil {
-			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
-		}
+		http.Error(w, "'latest' is not a valid value for 'gitRef'. Please use an immutable image.", http.StatusBadRequest)
+		return
 	}
 
 	cacheKey := fmt.Sprintf("%s:%s", repo, gitRef)
 
-	// Check cache
+	// Check cache first
 	if cachedResponse, ok := deps.getFromCache(cacheKey); ok {
-		w.WriteHeader(http.StatusOK)
-		_, _ = w.Write(cachedResponse)
-		return
-	}
-
-	// Determine if gitRef looks like a commit hash
-	commitHashRegex := regexp.MustCompile(`^[a-fA-F0-9]{7,40}$`)
-	if commitHashRegex.MatchString(gitRef) {
-		rec := &responseRecorder{ResponseWriter: w, statusCode: 200}
-		deps.CommitsHandler(rec, r)
-		deps.storeInCache(cacheKey, rec.body.Bytes()) // Cache response
+		w.WriteHeader(cachedResponse.StatusCode)
+		_, _ = w.Write(cachedResponse.Body)
 		return
 	}
 
 	// Try handling as a release first
-	releaseRecorder := &responseRecorder{ResponseWriter: w, statusCode: 200}
+	releaseRecorder := &responseRecorder{
+		ResponseWriter: httptest.NewRecorder(),
+		statusCode:     0, // Use a separate ResponseRecorder
+	}
 	deps.ReleasesHandler(releaseRecorder, r)
 
-	// If release is found, cache and return it
+	// If release is found (not 404), cache and return
 	if releaseRecorder.statusCode != http.StatusNotFound {
-		deps.storeInCache(cacheKey, releaseRecorder.body.Bytes())
+		w.WriteHeader(releaseRecorder.statusCode)
+		if _, err := w.Write(releaseRecorder.body.Bytes()); err != nil {
+			log.Printf("Error writing response: %v", err)
+		}
+		deps.storeInCache(cacheKey, releaseRecorder.statusCode, releaseRecorder.body.Bytes())
 		return
 	}
 
-	// Fall back to commits if no release is found
-	commitRecorder := &responseRecorder{ResponseWriter: w, statusCode: 200}
-	deps.CommitsHandler(commitRecorder, r)
+	// Clear previous response before falling back
+	w.Header().Del("Content-Type") // Ensure correct content type is set by CommitsHandler
+	w.WriteHeader(http.StatusOK)   // Reset status before falling back
 
-	// Cache the final commit response
-	deps.storeInCache(cacheKey, commitRecorder.body.Bytes())
+	// Fall back to CommitsHandler
+	deps.processAndCacheResponse(deps.CommitsHandler, w, r, cacheKey)
 }
 
-func (deps *HandlerDeps) getFromCache(key string) ([]byte, bool) {
+func (deps *HandlerDeps) processAndCacheResponse(
+	handler http.HandlerFunc,
+	w http.ResponseWriter,
+	r *http.Request,
+	cacheKey string,
+) {
+	rec := &responseRecorder{ResponseWriter: w, statusCode: 0} // Use 0 so that we can catch missing status updates
+	handler(rec, r)
+
+	// Store both status and response body
+	deps.storeInCache(cacheKey, rec.statusCode, rec.body.Bytes())
+}
+
+// Expiration durations
+const (
+	SuccessCacheDuration = 24 * time.Hour // 24 hours for 200 responses
+	ErrorCacheDuration   = 1 * time.Hour  // 1 hour for non-200 responses
```

Comment:
Should we make this cache duration configurable, maybe by some environment variable?